### DISCUSSION TITLE
rpmsg_char on Linux and openamp on baremetal with multiple endpoints

### DISCUSSION THREAD

**Discussion by yannik93** [OpenAMP/open-amp#279] [Q&A]
Hi,
iam trying to implement multiple Endpoints on Linux and Baremtetal side. 
On Linux side iam using the rpmsg_char driver and on Baremetal just the basic OpenAMP.
I can establish and send data via one Endpoint between my Linux and Baremtal, but actually i cant exchange data via a second created endpoint.  Does anyone worked already with multiple endpoints with the rpmsg_char driver and could give me some help.

**Comment by arnopo** [comment]
Hi @yannik93 
I guess this is the continuation of our discussion started here:
https://community.st.com/s/question/0D73W000000nIU2/detail?s1oid=00Db0000000YtG6&s1nid=0DB0X000000DYbd&emkind=chatterCommentNotification&s1uid=0050X000008AiF8&emtm=1617011093989&fromEmail=1&s1ext=0
Could you confirm?

In this case the code you run on Linux is not generic so if you need support from community, please provide links to: 
- Linux kernel source code 
- Linux Application source code
- Copro application code

**Reply by yannik93** [reply]
Hi,
yes you are right about the discussion. 
Anyways i didnt use any patch for the rpmsg_char driver so far. I just enabled the rpmsg_char driver via yocto with bitbake virtual/kernel -c menuconfig to my own distribution based on the weston image and it works fine for me with one endpoint/channel but not with multiple. 
I attached the code of the Copro (which is based on the OpenAMP_raw example) and from the Linux side.
I dont even know if i have the right understanding of endpoints and channels. What i need is just two "Channels", one Channel which i just wanna use to exchange data and one channel to exchange informations. And for that i wanna use the rpmsg_char driver, because i get dataloss with the tty_driver.    

[code.zip](https://github.com/OpenAMP/open-amp/files/6236856/code.zip)

**Comment by arnopo** [comment]
Ok i understood now how you probe it, using bind mechanism . But regarding your code i don't understand how it can work even for one endpoint. 
I tested your code, you have 2 endpoints created on copro side: ept address 0 and ept address 1
on Linux side you have also 2 endpoints but with source and destinations address set to -1 
with this settings, seems that some messages are sent from a Linux enpoint address 0x400 to the coprocessor endpoint 0.

There are 2 issues in you implementation:

1) The rpmsg_char driver is a specific driver that can be instantiate only one time per coprocessor to create the /sys/class/rpmsg/rpmsg_ctrlX device. So bind it only one time. 

2)  Then you have to create 2 endpoints with the /sys/class/rpmsg/rpmsg_ctrl0,  but by specifying source and destination addresses
 For your exemple :
- endpoint 1: src = 0x100 ( or any value different from -1)  and dst = 0x0 (the address of the first coprocessor endpoint )
- endpoint 2: src = 0x101 ( or any value different from -1) and dst = 0x01 (the address of the second coprocessor endpoint)
 
But i assume that without the series  https://patchwork.kernel.org/project/linux-remoteproc/list/?series=446305&state=%2A&archive=both, it still wont work, you will probably have some endpoint address issues.
So please first backport this series (that should be integrated in next kernel release), and then remove the bind part of your code as the rpmsg_char device will be automatically bound when you will start the coprocessor.

**Reply by yannik93** [reply]
Hi,
first thank you for the fast reply.
I just made test with your notes and you are right, it doesnt work (for mutliple endpoints) without the patches.
Anyways its good to know that in the next release this limitation wont be there anymore. 
Actually its not so important to get it work for me, because iam just doing my masterthesis about evaluating the OpenAMP-Framework, so its good to have this information.
If i have still some time i will try the patches.

Thank you so far.
